### PR TITLE
Mtn

### PR DESCRIPTION
**Pull request by dkarpukhin**

### CODE REVIEW

File: daemon/daemon.go [hunk 1]
```diff
@@ -30,6 +33,15 @@ func New(ctx context.Context, configuration *config.Config) (*Daemon, error) {
 		cfg:       configuration,
 		boxes:     make(isolate.Boxes),
 		listeners: make([]net.Listener, 0),
+		state: make(globalState)
```
Comment (noxiouz) [inline daemon/daemon.go]: ,

File: daemon/daemon.go [hunk 2]
```diff
@@ -30,6 +33,15 @@ func New(ctx context.Context, configuration *config.Config) (*Daemon, error) {
...
+	}
+
+	d.state["mtn"] = mtnState{}
+	if !d.state["mtn"].cfgInit(configuration) {
```
Comment (noxiouz) [inline daemon/daemon.go]: interface{} has no method cfgInit

File: daemon/daemon.go [hunk 3]
```diff
@@ -30,6 +33,15 @@ func New(ctx context.Context, configuration *config.Config) (*Daemon, error) {
...
+		return nil, fmt.Errorf("cant do cfgInit() for MTN with configuration: %s", configuration)
+	}
+	if !d.state["mtn"].poolInit() {
```
Comment (noxiouz) [inline daemon/daemon.go]: see above

File: isolate/mtn.go [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+		c.cfg.ident = c.cfg.schedLabel
+	} else {
+		c.cfg.ident = cfg.Mtn.Ident
+	}
+	c.cfg.url = cfg.Mtn.Url
+	c.cfg.headers = cfg.Mtn.Headers
+	return true
+}
+
+func (c *mtnState) poolInit() bool {
```
Comment (noxiouz) [inline isolate/mtn.go]: why not return error?

File: isolate/mtn.go [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+		return true
+	}
+	c.pool = make(mtnPool)
+	all_allocs, err := c.getAllocations()
+	if err != nil {
+		//print err
+		return false
+	}
+	for net_id, allocs := range all_allocs {
+		c_state := *new(idState)
```
Comment (noxiouz) [inline isolate/mtn.go]: no need to deref pointer

File: isolate/mtn.go [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+		for a := range allocs {
+			c_state.allocations[a["id"]] = a
+		}
+		c.pool[net_id] = &c_state
+	}
+	return true
+}
+
+func (c *mtnState) getAllocations() map[string][]Allocation error {
+	ctx := context.WithTimeout(context.Background(), 20 * time.Second)
```
Comment (noxiouz) [inline isolate/mtn.go]: ctx, cancel :=

defer cancel()

File: isolate/mtn.go [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+	req = req.WithContext(ctx)
+	for i := 0; i < c.cfg.allocbuffer; i++ {
+		resp, do_err := http.DefaultClient.Do(req)
+		if do_err != nil {
+			return _, do_err
+		}
+		jresp := rawAlloc{}
+		unm_err := json.Unmarshal(resp, jresp)
+		if unm_err != nil {
+			return _, unm_err
```
Comment (noxiouz) [inline isolate/mtn.go]: nil, unm_err

File: isolate/mtn.go [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+			}
+			c.pool[netid].reserved += c.cfg.allocbuffer
+			for id, alloc := range allocs {
+				//maybe here need check for new id is not exist
+				c.pool[netid].allocations[id] = alloc
+			}
+			c.pool[netid].Unlock()
+			return true
+		}
+	} else {
```
Comment (noxiouz) [inline isolate/mtn.go]: no need to have else, if branches have return

File: isolate/porto/profile_decodable.go [hunk 9]
```diff
@@ -172,7 +199,16 @@ func (z *Profile) EncodeMsg(en *msgp.Writer) (err error) {
 	if err != nil {
 		return
 	}
-	// write "cwd"
+        // write "network"
+        err = en.Append(0xa7, 0x6e, 0x65, 0x74, 0x77, 0x6f, 0x72, 0x6b)
```
Comment (noxiouz) [inline isolate/porto/profile_decodable.go]: magic number?

**Review by noxiouz — CHANGES_REQUESTED** [review]

File: daemon/daemon.go (hunk 1 above)
Comment (dkarpukhin) [reply daemon/daemon.go]: Now, after 19153e5872faa3d53280a1f4b492d996bddc4d22 commit, compile with go1.8.3.

File: isolate/mtn.go [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+	Host string `json:"host"`
+	Scheduler string `json:"scheduler"`
+}
+
+func (c *MtnState) CfgInit(cfg *Config) bool {
+	c.Cfg.Enable = cfg.Mtn.Enable
+	if !c.Cfg.Enable {
+		return true
+	}
+	if cfg.Mtn.Allocbuffer < 1 {
```
Comment (noxiouz) [inline isolate/mtn.go]: == 0?

File: isolate/mtn.go [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+	if do_err != nil {
+		fmt.Println("GetAllocationsE2: ", do_err)
+		return nil, do_err
+	}
+	r := make(map[string][]Allocation)
+	jresp := []RawAlloc{}
+	fmt.Println("GetAllocations1: ", rh)
+	decoder := json.NewDecoder(rh.Body)
+	r_err := decoder.Decode(&jresp)
+	defer rh.Body.Close()
```
Comment (noxiouz) [inline isolate/mtn.go]: выше

File: isolate/mtn.go [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+	}
+	for _, a := range jresp {
+		r[a.Network] = append(r[a.Network], Allocation{a.Porto.Net, a.Porto.Hostname, a.Porto.Ip, a.Id, false})
+	}
+	return r, nil
+}
+
+func (c *MtnState) RequestAllocs(netid string) (map[string]Allocation, error) {
+	r := make(map[string]Allocation)
+	ctx, _ := context.WithTimeout(context.Background(), 20 * time.Second)
```
Comment (noxiouz) [inline isolate/mtn.go]: defer cancel()

File: isolate/mtn.go [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+		rh, do_err := http.DefaultClient.Do(req)
+		if do_err != nil {
+			fmt.Println("DEBUG_RequestAllocs2En: ", do_err)
+			return nil, do_err
+		}
+		jresp := RawAlloc{}
+	        decoder := json.NewDecoder(rh.Body)
+		r_err := decoder.Decode(&jresp)
+		fmt.Println("DEBUG_RequestAllocs2jresp: ", jresp)
+		defer rh.Body.Close()
```
Comment (noxiouz) [inline isolate/mtn.go]: no defer тогда уж

**Comment by dkarpukhin** [comment]
Clear some dirty parts and compile now.

File: isolate/mtn.go [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+	}
+	c.Pool = make(map[string]*IdState)
+	allAllocs, err := c.GetAllocations()
+	if err != nil {
+		// TODO: maybe need use main logger at this step
+		fmt.Println("%s ERROR: Cant init pool inside PoolInit(), err: %s", time.Now().UTC().Format(time.RFC3339), err)
+		return false
+	}
+	for netId, allocs := range allAllocs {
+		cState := *new(IdState)
```
Comment (noxiouz) [inline isolate/mtn.go]: ? why deref of pointer? just var cState IdState
or better cState := IdState{
Reserverd: 0,
Allocation: ....
} etc

File: isolate/mtn.go [hunk 15]
```diff
@@ -0,0 +1,294 @@
...
+	req, nrErr := http.NewRequest("GET", c.Cfg.Url + "?scheduler=" + c.Cfg.SchedLabel, nil)
+	if nrErr != nil {
+		return nil, nrErr
+	}
+	for header, value := range c.Cfg.Headers {
+		req.Header.Set(header, value)
+	}
+	req = req.WithContext(ctx)
+	rh, doErr := http.DefaultClient.Do(req)
+	defer rh.Body.Close()
```
Comment (noxiouz) [inline isolate/mtn.go]: check err before Body.Close

File: isolate/mtn.go [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+			rErr := decoder.Decode(&errResp)
+			if rErr != nil {
+				return nil, fmt.Errorf(
+					"%s ERROR: Cant allocate. Request error: %s. Internal error: %s. Caused: %s.",
+					time.Now().UTC().Format(time.RFC3339),
+					doErr,
+					errResp.Message[0],
+					errResp.Cause.Message[0],
+				)
+			} else {
```
Comment (noxiouz) [inline isolate/mtn.go]: redundant else

File: isolate/mtn.go [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+	r := make(map[string]Allocation)
+	ctx, cancel := context.WithTimeout(context.Background(), 20 * time.Second)
+	defer cancel()
+	jsonBody := PostAllocreq{netid, c.Cfg.Ident, c.Cfg.SchedLabel}
+	txtBody, mrshErr := json.Marshal(jsonBody)
+	if mrshErr != nil {
+		return nil, mrshErr
+	}
+	for i := 0; i < c.Cfg.Allocbuffer; i++ {
+		req, nrErr := http.NewRequest("POST", c.Cfg.Url, bytes.NewBuffer(txtBody))
```
Comment (noxiouz) [inline isolate/mtn.go]: bytes.NewReader

File: isolate/mtn.go [hunk 18]
```diff
@@ -0,0 +1,294 @@
...
+		if nrErr != nil {
+			return nil, nrErr
+		}
+		for header, value := range c.Cfg.Headers {
+			req.Header.Set(header, value)
+		}
+		req.Header.Set("Content-Type", "application/json")
+		req = req.WithContext(ctx)
+		rh, doErr := http.DefaultClient.Do(req)
+		defer rh.Body.Close()
```
Comment (noxiouz) [inline isolate/mtn.go]: err before defer and please, make sure u need defer inside a loop

File: isolate/mtn.go [hunk 19]
```diff
@@ -0,0 +1,294 @@
...
+			return reqErr
+		}
+		newPool := IdState{*new(sync.Mutex), c.Cfg.Allocbuffer, allocs}
+		c.Pool[netId] = &newPool
+		return nil
+	}
+}
+
+func (c *MtnState) UnbindAllocs(netId string) {
+	c.Pool[netId].Lock()
```
Comment (noxiouz) [inline isolate/mtn.go]: Is this method called under c.Pool level Lock

File: isolate/mtn.go [hunk 20]
```diff
@@ -0,0 +1,294 @@
...
+	c.Pool[netId].Reserved -= c.Cfg.Allocbuffer
+	c.Pool[netId].Unlock()
+}
+
+func (c *MtnState) UseAlloc(netId string) (Allocation, error) {
+	c.Pool[netId].Lock()
```
Comment (noxiouz) [inline isolate/mtn.go]: See comment in UnbindAllocs

File: isolate/mtn.go [hunk 21]
```diff
@@ -0,0 +1,294 @@
...
+}
+
+func (c *MtnState) UnuseAlloc(netId string, id string) {
+	c.Pool[netId].Lock()
+	newPool := make(map[string]*IdState)
+	newAllocs := make(map[string]Allocation)
+	for cId, alloc := range c.Pool[netId].Allocations {
+		if cId == id {
+			alloc.Used = false
+			newAllocs[id] = alloc
```
Comment (noxiouz) [inline isolate/mtn.go:284]: newAlloc is not used

File: isolate/mtn.go [hunk 22]
```diff
@@ -0,0 +1,294 @@
...
+func (c *MtnState) UnbindAllocs(netId string) {
+	c.Pool[netId].Lock()
+	c.Pool[netId].Reserved -= c.Cfg.Allocbuffer
+	c.Pool[netId].Unlock()
+}
+
+func (c *MtnState) UseAlloc(netId string) (Allocation, error) {
+	c.Pool[netId].Lock()
+	newallocs := make(map[string]Allocation)
+	finded := false
```
Comment (noxiouz) [inline isolate/mtn.go]: found

File: isolate/mtn.go (hunk 20 above)
Comment (noxiouz) [inline isolate/mtn.go]: No Unlock?

File: isolate/mtn.go [hunk 23]
```diff
@@ -0,0 +1,294 @@
...
+	c.Pool[netId].Lock()
+	newPool := make(map[string]*IdState)
+	newAllocs := make(map[string]Allocation)
+	for cId, alloc := range c.Pool[netId].Allocations {
+		if cId == id {
+			alloc.Used = false
+			newAllocs[id] = alloc
+		}
+	}
+	newPool[netId].Unlock()
```
Comment (noxiouz) [inline isolate/mtn.go]: do u realy want to unlock this?

File: isolate/mtn.go (hunk 14 above)
Comment (dkarpukhin) [reply isolate/mtn.go]: fixed, tnx

File: isolate/mtn.go (hunk 15 above)
Comment (dkarpukhin) [reply isolate/mtn.go]: fixed, tnx

File: isolate/mtn.go (hunk 16 above)
Comment (dkarpukhin) [reply isolate/mtn.go]: fixed, tnx

File: isolate/mtn.go (hunk 17 above)
Comment (dkarpukhin) [reply isolate/mtn.go]: fixed, tnx

File: isolate/mtn.go (hunk 18 above)
Comment (dkarpukhin) [reply isolate/mtn.go]: change that part, tnx

File: isolate/mtn.go (hunk 19 above)
Comment (dkarpukhin) [reply isolate/mtn.go]: tnx, i tiny rework locks usage now

File: isolate/mtn.go (hunk 20 above)
Comment (dkarpukhin) [reply isolate/mtn.go]: tnx, i tiny rework locks usage now

File: isolate/mtn.go (hunk 22 above)
Comment (dkarpukhin) [reply isolate/mtn.go]: fixed, tnx

File: isolate/mtn.go (hunk 21 above)
Comment (dkarpukhin) [reply isolate/mtn.go:284]: fixed, tnx

File: isolate/mtn.go (hunk 20 above)
Comment (dkarpukhin) [reply isolate/mtn.go]: tnx, i tiny rework locks usage now

File: isolate/mtn.go (hunk 23 above)
Comment (dkarpukhin) [reply isolate/mtn.go]: tnx, i tiny rework locks usage now

**Pull request merged**